pkg/deploy/controller/deploymentconfig/controller_test.go

```diff
@@ -1,835 +1,629 @@
-			if err == nil {
-				t.Fatalf("expected error")
-			}
-			if reflect.TypeOf(err) != scenario.errorType {
-				t.Fatalf("error expected: %s, got: %s", scenario.errorType, reflect.TypeOf(err))
-			}
-		}
-
-		expectedCancellations := []int{}
-		actualCancellations := []int{}
-		for _, e := range scenario.existing {
-			if e.shouldCancel {
-				expectedCancellations = append(expectedCancellations, e.version)
-			}
-		}
-		for _, d := range updatedDeployments {
-			actualCancellations = append(actualCancellations, deployutil.DeploymentVersionFor(&d))
-		}
-
-		sort.Ints(actualCancellations)
-		sort.Ints(expectedCancellations)
-		if !reflect.DeepEqual(actualCancellations, expectedCancellations) {
-			t.Fatalf("expected cancellations: %v, actual: %v", expectedCancellations, actualCancellations)
-		}
-	}
-}
-
-func TestFindDetails(t *testing.T) {
-	existingDeployments := &kapi.ReplicationControllerList{}
-	controller := &DeploymentConfigController{
-		deploymentClient: &deploymentClientImpl{
-			listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
-				return existingDeployments, nil
-			},
+		{
+			name:             "scale up active (not latest) completed deployment",
+			replicas:         5,
+			newVersion:       2,
+			expectedReplicas: 5,
+			before: []deployment{
+				{version: 1, replicas: 1, replicasA: newint(1), status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
+			},
+			after: []deployment{
+				{version: 1, replicas: 5, replicasA: newint(5), status: deployapi.DeploymentStatusComplete, cancelled: false},
+				{version: 2, replicas: 0, replicasA: newint(0), desiredA: newint(1), status: deployapi.DeploymentStatusFailed, cancelled: true},
+			},
+			errExpected: false,
 		},
-		osClient: testclient.NewSimpleFake(),
-	}
-
-	type reaction struct {
-		verb, resource string
-		fn             ktestclient.ReactionFunc
-	}
-
-	tests := []struct {
-		name              string
-		controller        *DeploymentConfigController
-		version           int
-		hasDeployments    bool
-		hasNoImageChange  bool
-		hasMultipleErrors bool
-		status            deployapi.DeploymentStatus
-		reactions         []reaction
-		expectedDetails   string
-		expectedLatest    bool
-		expectedErr       bool
-	}{
 		{
-			name: "cannot get existing deployments",
-			controller: &DeploymentConfigController{
-				deploymentClient: &deploymentClientImpl{
-					listDeploymentsForConfigFunc: func(namespace, configName string) (*kapi.ReplicationControllerList, error) {
-						return nil, errors.New("cannot return these deployments")
-					},
-				},
-			},
-			version:         0,
-			hasDeployments:  false,
-			expectedDetails: "",
-			expectedLatest:  false,
-			expectedErr:     true,
+			name:             "scale down latest/active completed deployment",
+			replicas:         1,
```

Comment:
Since we have conditions that deal with 0 as a special case, should we add in somes tests for scaling down to 0 in various conditions?